### ISSUE TITLE
Problem with String 'Null' and Postgres string arrays

### ISSUE THREAD

**Issue by ascastro** [rails/rails#22490]
So that it is easier to replicate this issue (DB is Postgres), let's assume we have the following migration:
```ruby
class CreatePeople < ActiveRecord::Migration
  def change
    create_table :people do |t|
      t.string :names, array: true, default: []

      t.timestamps null: false
    end
  end
end
```
and the model
```ruby
class Person < ActiveRecord::Base
end
```

Saving the string 'Null' (which can be a surname) to the array `names`, will cause the object to have a `nil` entry in the array:
```ruby
$ bin/rails console
Loading development environment (Rails 4.2.2)
> person = Person.create(names: ['Joe', 'Null'])
> person.reload.names
=> ["Joe", nil]
```

**Comment by denniss** [comment]
Failing test

```ruby
require 'active_record'
require 'minitest/autorun'
require 'logger'

begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'rails')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :people do |t|
    t.string :names, array: true, default: []
    t.timestamps null: false
  end
end

class Person < ActiveRecord::Base
end

class NullTest < Minitest::Test
  def test_null_element_array
    person = Person.create(names: ['Joe', 'Null'])
    assert_equal ['Joe', 'Null'], person.reload.names
  end
end
```

**Comment by sgrif** [comment]
Yeah, this has been fixed on master. Unfortunately, the fix isn't something we can backport to 4.2, as it relies on a lot of larger changes that were made to the PG infrastructure.
